Improve display of wide-layout attachment on iOS

squelart opened the pull request:
#### ed1422596dce5ff012e64a38faf402ac1674fc7e
<pre>
Improve display of wide-layout attachment on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=254142">https://bugs.webkit.org/show_bug.cgi?id=254142</a>
rdar://problem/106925105

Reviewed by Aditya Keerthi.

Progress towards the target specs, in particular the colors and icon display.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateThumbnail):
(WebCore::HTMLAttachmentElement::updateIcon):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
(div#attachment-container::selection):
(div#attachment-preview-area):
(attachment#attachment-preview):
(div#attachment-information-area):
(div#attachment-information-block):
(div#attachment-action):
(div#attachment-title):
(div#attachment-subtitle):
(button#attachment-save-button):
(div#attachment-save-icon):
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::AttachmentLayout):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::attachmentIntrinsicSize const):
(WebCore::RenderThemeIOS::paintAttachment):

Canonical link: <a href="https://commits.webkit.org/261905@main">https://commits.webkit.org/261905@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5422a67a40a513d615765b5a7eb628a01f81dfb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113181 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22314 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1854 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4956 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121643 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117309 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23686 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13494 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6194 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118968 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100851 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/106305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1392 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46633 "layout-tests (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14615 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1433 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/98882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15327 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20619 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53423 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8319 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17176 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -557,13 +557,19 @@ void HTMLAttachmentElement::updateThumbnail(const RefPtr<Image>& thumbnail)
     m_thumbnail = thumbnail;
     removeAttribute(HTMLNames::progressAttr);
+    if (m_innerLegacyAttachment)
+        m_innerLegacyAttachment->m_thumbnail = thumbnail;
```
Nit: Can we move this above `removeAttribute` and close to the other assignment?

squelart replied:
And I've just noticed I should actually call the `updateThumbnail` method instead, and same with `updateIcon` below.

pxlcoder commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -76,8 +99,13 @@ div#attachment-information-area {
     grid-column: 1;
+#if (defined(WTF_PLATFORM_IOS_FAMILY) && WTF_PLATFORM_IOS_FAMILY)
+    margin-left: 6px;
+    margin-right: 26px;
```
I missed this in earlier reviews, but this should probably use `margin-inline-start` / `margin-inline-end`, in order to support RTL languages.

I would recommend testing RTL, and then a followup change, since there are probably other places that need to change too.

squelart replied:
Follow-up: rdar://106966312 (Wide-layout attachment: Verify that RTL display is correct)

pxlcoder commented on Source/WebCore/rendering/AttachmentLayout.mm:
```diff
@@ -199,6 +199,7 @@ static Color titleTextColorForAttachment(const RenderAttachment& attachment, Att
 #if PLATFORM(IOS_FAMILY)
 
 constexpr CGSize attachmentSize = { 160, 119 };
+constexpr CGSize attachmentIconOnlySize = { 56, 72 };
```
Should probably leave a comment here and in the CSS file if the values are co-dependent.

pxlcoder commented on Source/WebCore/rendering/AttachmentLayout.mm:
```diff
@@ -278,6 +279,24 @@ static CGFloat attachmentDynamicTypeScaleFactor()
+
+        hasProgress = getAttachmentProgress(attachment, progress);
+        if (hasProgress)
+            progressRect = FloatRect((attachmentRect.width() / 2) - (attachmentProgressSize / 2), (attachmentRect.height() / 2) - (attachmentProgressSize / 2), attachmentProgressSize, attachmentProgressSize);
```
Nit: This could use `(attachmentRect.width() - attachmentProgressSize) / 2` and `(attachmentRect.height() - attachmentProgressSize) / 2`.

squelart replied:
I copied it from the existing code below! Maybe it was important to do the truncating divisions first? (Though I doubt it matters much, it may move by 1 pixel here&there.)
I'll change it only here in this new code.

pxlcoder replied:
Oh! Maybe just keep it the same then...

squelart committed:
Improve display of wide-layout attachment on iOS
https://bugs.webkit.org/show_bug.cgi?id=254142
rdar://problem/106925105

Reviewed by Aditya Keerthi.

Progress towards the target specs, in particular the colors and icon display.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateThumbnail):
(WebCore::HTMLAttachmentElement::updateIcon):
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
(div#attachment-container::selection):
(div#attachment-preview-area):
(attachment#attachment-preview):
(div#attachment-information-area):
(div#attachment-information-block):
(div#attachment-action):
(div#attachment-title):
(div#attachment-subtitle):
(button#attachment-save-button):
(div#attachment-save-icon):
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::AttachmentLayout):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::attachmentIntrinsicSize const):
(WebCore::RenderThemeIOS::paintAttachment):

Canonical link: https://commits.webkit.org/261905@main
